daffodil-runtime2/src/main/resources/org/apache/daffodil/runtime2/c/libcli/xml_writer.c

```diff
@@ -22,30 +22,55 @@
 #include <stdbool.h>     // for bool, false, true
 #include <stdint.h>      // for uint8_t, int16_t, int32_t, int64_t, int8_t, uint16_t, uint32_t, uint64_t
 #include <stdlib.h>      // for free, malloc
-#include <string.h>      // for strcmp
+#include <string.h>      // for strcmp, strlen
 #include "cli_errors.h"  // for CLI_XML_DECL, CLI_XML_ELEMENT, CLI_XML_WRITE, LIMIT_XML_NESTING
 #include "errors.h"      // for Error, Error::(anonymous)
 #include "stack.h"       // for stack_is_empty, stack_pop, stack_push, stack_top, stack_init
 // clang-format on
 
-// Fix a real number to conform to xsd:float syntax if needed
+// Fix a real number's syntax to remove as many textual differences
+// between runtime1/runtime2 as possible.  A better approach may be
+// to make TDMLRunner compare floats and doubles in XML numerically
+// with a small epsilon instead of textually character by character.
```

Comment:
The TDML Runner uses the XML comparison logic XMLUtils.scala, with the type aware comparison happening in the `textIsSame` function: 

https://github.com/apache/daffodil/blob/main/daffodil-lib/src/main/scala/org/apache/daffodil/xml/XMLUtils.scala#L1194-L1214=

Unfortunately, the way it knows how to use special comparison logic is with the `xsi:type` attribute, which must be on either the actual infoset or the expected infoset. We've been putting them on the actual infoset when we need the type aware comparison since the runtimes don't current support it (see [DAFFODIL-2402]](https://issues.apache.org/jira/browse/DAFFODIL-2402)). Though, it might be a bit of a headache to add it for all doubles/floats--we don't need this very often.

If this makes Java and C mostly the same, maybe this is a better approach. I'm also not against change Java to match C if that's easier.